--- src/stunnel.c.orig	2009-07-18 01:19:14.000000000 +0200
+++ src/stunnel.c	2009-07-18 01:19:18.000000000 +0200
@@ -61,7 +61,7 @@
 #ifndef USE_WIN32
 int main(int argc, char* argv[]) { /* execution begins here 8-) */
 
-    main_initialize(argc>1 ? argv[1] : NULL, argc>2 ? argv[2] : NULL);
+    main_initialize(argc>1 ? argv[1] : NULL, argc>2 ? argv[2] : NULL, argc>3 ? argv[3] : NULL);
 
     signal(SIGPIPE, SIG_IGN); /* avoid 'broken pipe' signal */
     if(signal(SIGTERM, SIG_IGN)!=SIG_IGN)
@@ -80,10 +80,10 @@
 }
 #endif
 
-void main_initialize(char *arg1, char *arg2) {
+void main_initialize(char *arg1, char *arg2, char *arg3) {
     ssl_init(); /* initialize SSL library */
     sthreads_init(); /* initialize critical sections & SSL callbacks */
-    parse_config(arg1, arg2);
+    parse_config(arg1, arg2, arg3);
 
 #ifdef USE_FIPS
     if(options.option.fips) {
--- src/options.c.orig	2009-07-18 00:55:30.000000000 +0200
+++ src/options.c	2009-07-18 01:20:04.000000000 +0200
@@ -1352,26 +1352,28 @@
         "[<filename>] ] "
 #ifndef USE_WIN32
         "-fd <n> "
+		"-pid <pidfile> "
 #endif
         "| -help | -version | -sockets");
-    s_log(LOG_RAW, "    <filename>  - use specified config file instead of %s",
+    s_log(LOG_RAW, "    <filename>      - use specified config file instead of %s",
         confname);
 #ifdef USE_WIN32
 #ifndef _WIN32_WCE
-    s_log(LOG_RAW, "    -install    - install NT service");
-    s_log(LOG_RAW, "    -uninstall  - uninstall NT service");
+    s_log(LOG_RAW, "    -install        - install NT service");
+    s_log(LOG_RAW, "    -uninstall      - uninstall NT service");
 #endif
-    s_log(LOG_RAW, "    -quiet      - don't display a message box on success");
+    s_log(LOG_RAW, "    -quiet          - don't display a message box on success");
 #else
-    s_log(LOG_RAW, "    -fd <n>     - read the config file from a file descriptor");
+    s_log(LOG_RAW, "    -fd <n>         - read the config file from a file descriptor");
+    s_log(LOG_RAW, "    -pid <pidfile>  - write pidfile there");
 #endif
-    s_log(LOG_RAW, "    -help       - get config file help");
-    s_log(LOG_RAW, "    -version    - display version and defaults");
-    s_log(LOG_RAW, "    -sockets    - display default socket options");
+    s_log(LOG_RAW, "    -help           - get config file help");
+    s_log(LOG_RAW, "    -version        - display version and defaults");
+    s_log(LOG_RAW, "    -sockets        - display default socket options");
     die(1);
 }
 
-void parse_config(char *name, char *parameter) {
+void parse_config(char *name, char *parameter, char *parameter2) {
 #ifdef CONFDIR
     char *default_config_file=CONFDIR CONFSEPARATOR "stunnel.conf";
 #else
@@ -1412,6 +1414,13 @@
         die(1);
     }
 #ifndef USE_WIN32
+    if(!strcasecmp(parameter, "-pid")) {
+        if(!parameter2) {
+            s_log(LOG_RAW, "No pid file specified");
+            syntax(default_config_file);
+        }
+        global_options(CMD_EXEC, "pid", parameter2);
+    }
     if(!strcasecmp(name, "-fd")) {
         if(!parameter) {
             s_log(LOG_RAW, "No file descriptor specified");
--- src/prototypes.h.orig	2009-07-18 01:23:37.000000000 +0200
+++ src/prototypes.h	2009-07-18 01:26:51.000000000 +0200
@@ -71,7 +71,7 @@
 
 extern volatile int num_clients;
 
-void main_initialize(char *, char *);
+void main_initialize(char *, char *, char *);
 void main_execute(void);
 #if !defined (USE_WIN32) && !defined (__vms) && !defined(USE_OS2)
 void drop_privileges(void);
@@ -263,7 +263,7 @@
     OPT_UNION *opt_val[3];
 } SOCK_OPT;
 
-void parse_config(char *, char *);
+void parse_config(char *, char *, char *);
 
 /**************************************** Prototypes for ctx.c */
 
