diff -ruN drbd-utils-8.9.6.orig/user/shared/libgenl.h drbd-utils-8.9.6/user/shared/libgenl.h
--- drbd-utils-8.9.6.orig/user/shared/libgenl.h
+++ drbd-utils-8.9.6/user/shared/libgenl.h
@@ -851,7 +851,7 @@ static inline struct nlattr *nla_nest_st
 {
 	struct nlattr *start = (struct nlattr *)msg->tail;
 
-	if (nla_put(msg, attrtype, 0, NULL) < 0)
+	if (nla_put(msg, attrtype | NLA_F_NESTED, 0, NULL) < 0)
 		return NULL;
 
 	return start;
