diff -Nur nagircbot-0.0.29b.orig/anna.cpp nagircbot-0.0.29b/anna.cpp
--- nagircbot-0.0.29b.orig/anna.cpp	2009-05-24 16:56:28.000000000 +0200
+++ nagircbot-0.0.29b/anna.cpp	2008-06-26 21:00:25.000000000 +0200
@@ -38,8 +38,8 @@
 int default_sleep = 2;		/* initial sleep time between connect failures */
 int max_sleep_time = 300;
 double sleep_multiply_factor = 2.0;
-int minimum_time_for_successfull_login = 25; // one needs to be on-channel for at least 5 seconds to be considered a successfull login
-int join_timeout = 5;	// it should take no longer then 5 seconds to join a channel, otherwhise: abort connection and retry
+int minimum_time_for_successfull_login = 25; // one needs to be on-channel for at least 5 seconds to be considered a successfull login
+int join_timeout = 15;	// it should take no longer then 5 seconds to join a channel, otherwhise: abort connection and retry
 int max_n_join_tries = 2;	// try 2 times to get on a channel
 char *server = "localhost:6667";	/* default irc server */
 char *channel = "#nagircbot";	/* default channel to connect to */
