--- nnrpd/python.c.orig	2010-10-31 02:12:41.000000000 +0100
+++ nnrpd/python.c	2010-10-31 02:14:21.000000000 +0100
@@ -250,6 +250,7 @@
     char        *buffer;
     int		authnum;
     int		i;
+    Py_ssize_t  pos;
 
     PY_load_python();
     proc = PY_setup(PYTHONaccess, PYTHONmain, file);
@@ -312,17 +313,17 @@
     vector_resize(access_vec, PyDict_Size(result) - 1);
 
     /* Store dict values in proper format in access vector. */
-    i = 0;
+    pos = 0;
     buffer = xmalloc(BIG_BUFFER);
 
-    while(PyDict_Next(result, &i, &key, &value)) {
+    while(PyDict_Next(result, &pos, &key, &value)) {
         if (!PyString_Check(key)) {
-            syslog(L_ERROR, "python access method return dictionary key %i not a string", i);
+            syslog(L_ERROR, "python access method return dictionary key %i not a string", pos);
             Reply("%d Internal error (2).  Goodbye!\r\n", NNTP_FAIL_TERMINATING);
             ExitWithStats(1, false);
         }
         if (!PyString_Check(value)) {
-            syslog(L_ERROR, "python access method return dictionary value %i not a string", i);
+            syslog(L_ERROR, "python access method return dictionary value %i not a string", pos);
             Reply("%d Internal error (2).  Goodbye!\r\n", NNTP_FAIL_TERMINATING);
             ExitWithStats(1, false);
         }
