--- kgpg/keyinfowidget.cpp.orig	2009-07-02 14:33:15.000000000 +0200
+++ kgpg/keyinfowidget.cpp	2009-07-03 13:15:55.000000000 +0200
@@ -111,6 +111,7 @@
 FILE *pass;
 char line[200]="";
 QString gpgOutput,fullID;
+bool noID=false;
 
 QString gpgcmd="gpg --no-tty --no-secmem-warning --with-colon --with-fingerprint --list-key "+KShellProcess::quote(Keyid);
 
@@ -119,6 +120,36 @@
                 gpgOutput=QString::fromUtf8(line);
                 if (gpgOutput.startsWith("uat"))
                         hasPhoto=true;
+                if (noID && gpgOutput.startsWith("uid")) {
+                    noID=false;
+                    QString fullname=gpgOutput.section(':',9,9);
+                    if (fullname.find("<")!=-1) {
+                        QString kmail=fullname;
+                        if (fullname.find(")")!=-1)
+                            kmail=kmail.section(')',1);
+                        kmail=kmail.section('<',1);
+                        kmail.truncate(kmail.length()-1);
+                        if (kmail.find("<")!=-1) ////////  several email addresses in the same key
+                        {
+                            kmail=kmail.replace(">",";");
+                            kmail.remove("<");
+                        }
+                        prop->tLMail->setText("<qt><a href=mailto:"+kmail+">"+kmail+"</a></qt>");
+                    } else
+                        prop->tLMail->setText(i18n("none"));
+
+                    QString kname=fullname.section('<',0,0);
+                    if (fullname.find("(")!=-1) {
+                        kname=kname.section('(',0,0);
+                        QString comment=fullname.section('(',1,1);
+                        comment=comment.section(')',0,0);
+                        prop->tLComment->setText(KgpgInterface::checkForUtf8(comment));
+                    } else {
+                        prop->tLComment->setText(i18n("none"));
+                    }
+
+                    prop->tLName->setText("<qt><b>"+KgpgInterface::checkForUtf8(kname).replace(QRegExp("<"),"&lt;")+"</b></qt>");
+                }
                 if (gpgOutput.startsWith("pub")) {
                         QString algo=gpgOutput.section(':',3,3);
                         switch( algo.toInt() ) {
@@ -204,14 +235,16 @@
 
                         QString fullname=gpgOutput.section(':',9,9);
 
-                        QDate date = QDate::fromString(gpgOutput.section(':',5,5), Qt::ISODate);
-                        prop->tLCreation->setText(KGlobal::locale()->formatDate(date));
+                        QDateTime date;
+                        date.setTime_t(gpgOutput.section(':',5,5).toUInt());
+                        prop->tLCreation->setText(KGlobal::locale()->formatDate(date.date()));
 
-			if (gpgOutput.section(':',6,6).isEmpty()) expirationDate=i18n("Unlimited");
+			if (gpgOutput.section(':',6,6).isEmpty())
+                            expirationDate=i18n("Unlimited");
 			else
 			{
-			date = QDate::fromString(gpgOutput.section(':',6,6), Qt::ISODate);
-			expirationDate=KGlobal::locale()->formatDate(date);
+                            date.setTime_t(gpgOutput.section(':',6,6).toUInt());
+                            expirationDate=KGlobal::locale()->formatDate(date.date());
 			}
                         prop->tLExpiration->setText(expirationDate);
 
@@ -241,31 +274,35 @@
                         }
                         prop->kCOwnerTrust->setCurrentItem(ownerTrust);
 
-                        if (fullname.find("<")!=-1) {
-                                QString kmail=fullname;
-				if (fullname.find(")")!=-1)
-				kmail=kmail.section(')',1);
-				kmail=kmail.section('<',1);
-				kmail.truncate(kmail.length()-1);
-				if (kmail.find("<")!=-1) ////////  several email addresses in the same key
-				{
-				kmail=kmail.replace(">",";");
-				kmail.remove("<");
-				}
-				prop->tLMail->setText("<qt><a href=mailto:"+kmail+">"+kmail+"</a></qt>");
-                        } else
-                                prop->tLMail->setText(i18n("none"));
-
-                        QString kname=fullname.section('<',0,0);
-                        if (fullname.find("(")!=-1) {
-                                kname=kname.section('(',0,0);
-                                QString comment=fullname.section('(',1,1);
-                                comment=comment.section(')',0,0);
-                                prop->tLComment->setText(KgpgInterface::checkForUtf8(comment));
-                        } else
-                                prop->tLComment->setText(i18n("none"));
+                        if(fullname.isEmpty()) {
+                                noID=true;
+                        } else {
+                                if (fullname.find("<")!=-1) {
+                                        QString kmail=fullname;
+                                        if (fullname.find(")")!=-1)
+                                        kmail=kmail.section(')',1);
+                                        kmail=kmail.section('<',1);
+                                        kmail.truncate(kmail.length()-1);
+                                        if (kmail.find("<")!=-1) ////////  several email addresses in the same key
+                                        {
+                                        kmail=kmail.replace(">",";");
+                                        kmail.remove("<");
+                                        }
+                                        prop->tLMail->setText("<qt><a href=mailto:"+kmail+">"+kmail+"</a></qt>");
+                                } else
+                                        prop->tLMail->setText(i18n("none"));
+
+                                QString kname=fullname.section('<',0,0);
+                                if (fullname.find("(")!=-1) {
+                                        kname=kname.section('(',0,0);
+                                        QString comment=fullname.section('(',1,1);
+                                        comment=comment.section(')',0,0);
+                                        prop->tLComment->setText(KgpgInterface::checkForUtf8(comment));
+                                } else
+                                        prop->tLComment->setText(i18n("none"));
 
-			prop->tLName->setText("<qt><b>"+KgpgInterface::checkForUtf8(kname).replace(QRegExp("<"),"&lt;")+"</b></qt>");
+                                prop->tLName->setText("<qt><b>"+KgpgInterface::checkForUtf8(kname).replace(QRegExp("<"),"&lt;")+"</b></qt>");
+                        }
 
                 }
 		if (gpgOutput.startsWith("fpr") && (fingervalue.isNull())) {
--- kgpg/kgpgoptions.cpp.orig	2009-07-02 14:31:10.000000000 +0200
+++ kgpg/kgpgoptions.cpp	2009-07-03 10:19:18.000000000 +0200
@@ -532,20 +532,28 @@
         }
         pclose(fp2);
 
-
+        bool noID=false;
+        bool appendToKeyName=false;
         fp = popen("gpg --no-tty --with-colon --list-keys", "r");
         while ( fgets( line, sizeof(line), fp)) {
                 tst=line;
                 if (tst.startsWith("pub")) {
                         name=KgpgInterface::checkForUtf8(tst.section(':',9,9));
-                        if ((!name.isEmpty()) && (trustedvals.find(tst.section(':',1,1))==-1)) {
+                        if (trustedvals.find(tst.section(':',1,1))==-1) {
                                 counter++;
                                 //name=name.section('<',-1,-1);
                                 //  name=name.section('>',0,0);
-                                names+=name;
                                 ids+=tst.section(':',4,4);
-                                if (tst.section(':',4,4).right(8)==alwaysKeyID)
+                                if (tst.section(':',4,4).right(8)==alwaysKeyID) {
                                         alwaysKeyName=tst.section(':',4,4).right(8)+":"+name;
+                                        appendToKeyName=true;
+                                }       
+                                if (name.isEmpty()) {
+                                        noID=true;
+                                }else{
+                                        names+=name;
+                                        appendToKeyName=false;
+                                }
 				if (issec.find(tst.section(':',4,4).right(8),0,FALSE)!=-1)
 				{
 //***                           page1->file_key->insertItem(pixkeyDouble,tst.section(':',4,4).right(8)+":"+name);
@@ -557,6 +565,14 @@
 //***                           page1->always_key->insertItem(pixkeySingle,tst.section(':',4,4).right(8)+":"+name);
 				}
                         }
+                } else if(noID && tst.startsWith("uid")) {
+                        name=KgpgInterface::checkForUtf8(tst.section(':',9,9));
+                        names+=name;
+                        noID=false;
+                        if (appendToKeyName) {
+                                alwaysKeyName+=name;
+                                appendToKeyName=false;
+                        }
                 }
         }
         pclose(fp);
--- kgpg/listkeys.cpp.orig	2009-07-02 14:26:08.000000000 +0200
+++ kgpg/listkeys.cpp	2009-07-02 23:07:29.000000000 +0200
@@ -279,6 +279,8 @@
                                         if (tst2.section(':',11,11).find('D')!=-1)
                                                 dead=true;
 					break;
+                                } else if (tst2.startsWith("uid") && fullname.isEmpty()) {
+                                    fullname=tst2.section(':',9,9);
                                 }
                         }
                         pclose(fp2);
@@ -2556,7 +2558,7 @@
         kdDebug(2100)<<"Expanding Key: "<<item->text(6)<<endl;
 
         cycle="pub";
-        bool noID=false;
+        bool noID=true;
         fp = popen(QFile::encodeName(QString("gpg --no-secmem-warning --no-tty --with-colons --list-sigs "+item->text(6))), "r");
 
         while ( fgets( tmpline, sizeof(tmpline), fp)) {
@@ -2727,8 +2729,12 @@
 
                         if (openKeys.find(pubKey.gpgkeyid)!=-1)
                                 item->setOpen(true);
+                } else if (tst.startsWith("uid") && noID) {
+                    gpgKey pubKey=extractKey(tst);
+                    item->setText(0, pubKey.gpgkeyname);
+                    item->setText(1, pubKey.gpgkeymail);
+                    noID = false;
                 }
-
         }
         pclose(fp);
         if (!issec.isEmpty())
@@ -2779,6 +2785,7 @@
         char line[300];
         UpdateViewItem *item=NULL;
         bool keyFound=false;
+        bool noID=false;
         fp = popen("gpg --no-secmem-warning --no-tty --with-colons --list-secret-keys", "r");
         while ( fgets( line, sizeof(line), fp)) {
                 QString lineRead=QString::fromUtf8(line);
@@ -2791,11 +2798,16 @@
                         //               isbold=true;
                         if (orphanedKey.gpgkeytrust==i18n("Expired"))
                                 isexpired=true;
-                        //          if (orphanedKey.gpgkeyname.isEmpty())
-                        //              noID=true;
+                        if (orphanedKey.gpgkeyname.isEmpty())
+                                noID=true;
 
                         item=new UpdateViewItem(this,orphanedKey.gpgkeyname,orphanedKey.gpgkeymail,QString::null,orphanedKey.gpgkeyexpiration,orphanedKey.gpgkeysize,orphanedKey.gpgkeycreation,orphanedKey.gpgkeyid,isbold,isexpired);
                         item->setPixmap(0,pixkeyOrphan);
+                } else if(noID && lineRead.startsWith("uid")) {
+                    gpgKey uidKey = extractKey(lineRead);
+                    item->setText(0, uidKey.gpgkeyname);
+                    item->setText(1, uidKey.gpgkeymail);
+                    noID = false;
                 }
         }
         pclose(fp);
@@ -2813,6 +2825,8 @@
 {
         FILE *fp;
         char line[300];
+        bool noID=false;
+        UpdateViewItem *item=NULL;
         fp = popen("gpg --no-secmem-warning --no-tty --with-colons --list-secret-keys", "r");
         while ( fgets( line, sizeof(line), fp)) {
                 QString lineRead=QString::fromUtf8(line);
@@ -2825,11 +2839,16 @@
                         //               isbold=true;
                         if (orphanedKey.gpgkeytrust==i18n("Expired"))
                                 isexpired=true;
-                        //          if (orphanedKey.gpgkeyname.isEmpty())
-                        //              noID=true;
+                        if (orphanedKey.gpgkeyname.isEmpty())
+                                noID=true;
                         orphanList+=orphanedKey.gpgkeyid+",";
-                        UpdateViewItem *item=new UpdateViewItem(this,orphanedKey.gpgkeyname,orphanedKey.gpgkeymail,QString::null,orphanedKey.gpgkeyexpiration,orphanedKey.gpgkeysize,orphanedKey.gpgkeycreation,orphanedKey.gpgkeyid,isbold,isexpired);
+                        item=new UpdateViewItem(this,orphanedKey.gpgkeyname,orphanedKey.gpgkeymail,QString::null,orphanedKey.gpgkeyexpiration,orphanedKey.gpgkeysize,orphanedKey.gpgkeycreation,orphanedKey.gpgkeyid,isbold,isexpired);
                         item->setPixmap(0,pixkeyOrphan);
+                } else if(noID && lineRead.startsWith("uid") && item != NULL) {
+                    gpgKey uidKey = extractKey(lineRead);
+                    item->setText(0, uidKey.gpgkeyname);
+                    item->setText(1, uidKey.gpgkeymail);
+                    noID = false;
                 }
         }
         pclose(fp);
@@ -2941,6 +2960,7 @@
 
         QString tst;
         bool keyFound=false;
+        bool noID=false;
         QString cmd="gpg --no-secmem-warning --no-tty --with-colons --list-keys "+currentID;
         fp = popen(QFile::encodeName(cmd), "r");
         while ( fgets( line, sizeof(line), fp)) {
@@ -2954,6 +2974,8 @@
                                 isbold=true;
                         if (pubKey.gpgkeytrust==i18n("Expired"))
                                 isexpired=true;
+                        if (pubKey.gpgkeyname.isEmpty())
+                                noID=true;
                         item=new UpdateViewItem(this,pubKey.gpgkeyname,pubKey.gpgkeymail,QString::null,pubKey.gpgkeyexpiration,pubKey.gpgkeysize,pubKey.gpgkeycreation,pubKey.gpgkeyid,isbold,isexpired);
                         item->setPixmap(2,pubKey.trustpic);
 			item->setVisible(true);
@@ -2964,6 +2986,11 @@
                         } else {
                                 item->setPixmap(0,pixkeySingle);
                         }
+                } else if(noID && tst.startsWith("uid")) {
+                        gpgKey uidKey=extractKey(tst);
+                        item->setText(0,uidKey.gpgkeyname);
+                        item->setText(1,uidKey.gpgkeymail);
+                        noID=false;
                 }
         }
         pclose(fp);
@@ -3046,8 +3073,10 @@
         ret.gpgkeysize=keyString[2];
         ret.gpgkeycreation=keyString[5];
         if(!ret.gpgkeycreation.isEmpty()) {
-                QDate date = QDate::fromString(ret.gpgkeycreation, Qt::ISODate);
-                ret.gpgkeycreation=KGlobal::locale()->formatDate(date, true);
+//                QDate date = QDate::fromString(ret.gpgkeycreation, Qt::ISODate);
+                QDateTime date;
+                date.setTime_t(ret.gpgkeycreation.toUInt());
+                ret.gpgkeycreation=KGlobal::locale()->formatDate(date.date(), true);
         }
         QString tid=keyString[4];
         ret.gpgkeyid=QString("0x"+tid.right(8));
@@ -3066,10 +3095,9 @@
                 //ret.gpgkeyname=ret.gpgkeyname.section('(',0,0);
         } else {
                 ret.gpgkeymail=QString::null;
-		ret.gpgkeyname=fullname;
+                ret.gpgkeyname=fullname;
                 //ret.gpgkeyname=fullname.section('(',0,0);
         }
-
         //ret.gpgkeyname=KgpgInterface::checkForUtf8(ret.gpgkeyname); // FIXME lukas
 
         QString algo=keyString[3];
